stan/math/prim/prob/hmm_marginal_lpdf.hpp

```diff
@@ -0,0 +1,234 @@
+  int n_states = log_omegas.rows();
+  int n_transitions = log_omegas.cols() - 1;
+
+  alphas.col(0) = omegas.col(0).cwiseProduct(rho);
+
+  double norm = alphas.col(0).maxCoeff();
+  alphas.col(0) /= norm;
+  alpha_log_norms(0) = std::log(norm);
+
+  for (int n = 0; n < n_transitions; ++n) {
+    alphas.col(n + 1)
+        = omegas.col(n + 1).cwiseProduct(Gamma.transpose() * alphas.col(n));
+
+    double norm = alphas.col(n + 1).maxCoeff();
+    alphas.col(n + 1) /= norm;
+    alpha_log_norms(n + 1) = std::log(norm) + alpha_log_norms(n);
+  }
+
+  return log(alphas.col(n_transitions).sum()) + alpha_log_norms(n_transitions);
+}
+
+/**
+ * For a Hidden Markov Model with observation y, hidden state x,
+ * and parameters theta, return the log marginal density, log
+ * pi(y | theta). In this setting, the hidden states are discrete
+ * and take values over the finite space {1, ..., K}.
+ * The marginal lpdf is obtained via a forward pass, and
+ * the derivative is calculated with an adjoint method,
+ * see (Betancourt, Margossian, & Leos-Barajas, 2020).
+ *
+ * @tparam T_omega type of the log likelihood matrix
+ * @tparam T_Gamma type of the transition matrix
+ * @tparam T_rho type of the initial guess vector
+ *
+ * @param[in] log_omega log matrix of observational densities.
+ *              The (i, j)th entry corresponds to the
+ *              density of the ith observation, y_i,
+ *              given x_i = j.
+ * @param[in] Gamma transition density between hidden states.
+ *              The (i, j)th entry is the probability that x_n = j,
+ *              given x_{n - 1} = i. The rows of Gamma are simplexes.
+ * @param[in] rho initial state
+ * @throw <code>std::domain_error</code> if the rows of Gamma are
+ * not a simplex.
+ * @throw <code>std::invalid_argument</code> if the size of rho is not
+ * the number of states.
+ * @throw <code>std::domain_error</code> if rho is not a simplex.
+ * @return log marginal density.
+ */
+template <typename T_omega, typename T_Gamma, typename T_rho>
+inline return_type_t<T_omega, T_Gamma, T_rho> hmm_marginal_lpdf(
+    const Eigen::Matrix<T_omega, Eigen::Dynamic, Eigen::Dynamic>& log_omegas,
+    const Eigen::Matrix<T_Gamma, Eigen::Dynamic, Eigen::Dynamic>& Gamma,
+    const Eigen::Matrix<T_rho, Eigen::Dynamic, 1>& rho) {
+  int n_states = log_omegas.rows();
+  int n_transitions = log_omegas.cols() - 1;
+
+  check_square("hmm_marginal_lpdf", "Gamma", Gamma);
+  check_consistent_size("hmm_marginal_lpdf", "Gamma", row(Gamma, 1), n_states);
+  {
+    // Temporary vector to use check_simplex, which only works once
+    // column vectors.
+    Eigen::Matrix<T_Gamma, Eigen::Dynamic, Eigen::Dynamic> Gamma_transpose
+        = Gamma.transpose();
+    for (int i = 0; i < Gamma.cols(); i++) {
+      // CHECK -- does check_simplex not work on row-vectors?
+      check_simplex("hmm_marginal_lpdf", "Gamma[i, ]",
+                    col(Gamma_transpose, i + 1));
+    }
+  }
```

Comment:
[optional]
This is really inefficient as it's written as it involves both a transpose and pulling the columns out.
Add a `check_stochastic_matrix` function.  Or as a first step, generalize `check_simplex` so you don't need to transpose---it only requires some template fiddling.  

I also realized there's a huge performance bug in our complicated checks, and created a separate issue for it:   https://github.com/stan-dev/math/issues/1779

So if you do this, make sure not to make the same mistakes as this issue points out.

Reply:
Will raise a new issue and address this separately.

Reply:
Great.  Then you'll build that and use it for this PR so there's not a super-inefficient check_stochastic_matrix function.

And are you really having columns be the simplex?  I've never seen anyone handle HMMs that way, so that's also going to be super confusing for users.

Reply:
The condition checks that the rows of Gamma are a simplex. In the unit test:
```
  EXPECT_THROW_MSG(hmm_marginal_lpdf(log_omegas, Gamma_bad, rho),
                   std::domain_error,
                   "hmm_marginal_lpdf: Gamma[i, ] is not a valid simplex. "
                   "sum(Gamma[i, ]) = 2, but should be 1")
```

The code in the test checks the column of the transpose of Gamma (so the row of Gamma). This is because check_simplex only works on column vectors.

Reply:
Anyways, I'll create the `check_stochastic_matrix` function and we can merge it with this PR.